### PR TITLE
🥽using pooling model to restore git dependencies

### PR DESCRIPTION
**Pull request by OsmondJiang**
#3995 

Restore:
- Acquire a exclusive slot(id)
- Restoring(create or checkout work tree)
- Release exclusive slot(id) with restored flag

Build:
- Acquire a shared slot(id) with flag(restored == true)
- Building
- Release shared slot(id)

There are somethings we need keep in mind:

- Exclusive/Shared lock has a default expire time, which means the lock can't be held in one owner for too long(default is 6 hours) in case it crashed.

- Once slot is restored, there is a lockdown time for this slot to be used for another restore, in case the slot is just restored but not used for build yet.

### CODE REVIEW

File: src/docfx/restore/index/DependencySlotPool.cs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+                                    {
+                                        slot = i;
+                                        slot.Acquirer = acquirer;
+                                        break;
+                                    }
+                                }
+                            }
+                            break;
+                        default:
+                            throw new NotSupportedException($"{type} is not supported");
```
Comment (yufeih) [inline src/docfx/restore/index/DependencySlotPool.cs]: just remove `LockType.None`?

File: src/docfx/restore/index/DependencySlotPool.cs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+        }
+
+        public static List<T> GetSlots<T>(string restoreDir) where T : DependencySlot
+        {
+            Debug.Assert(!string.IsNullOrEmpty(restoreDir));
+
+            var slotsFile = Path.Combine(restoreDir, "index.json");
+            var content = File.Exists(slotsFile) ? File.ReadAllText(slotsFile) : string.Empty;
+
+            return JsonUtility.Deserialize<List<T>>(content) ?? new List<T>();
```
Comment (yufeih) [inline src/docfx/restore/index/DependencySlotPool.cs]: consider use a wrapper object like other data models, it is hard to extend an array.

File: src/docfx/restore/index/DependencyGitPool.cs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+                remote,
+                type,
+                slot =>
+                {
+                    // update branch and commit info to new rented slot
+                    slot.Commit = commit;
+                    slot.Branch = branch;
+                    return slot;
+                },
+                existingSlot => existingSlot.Branch == branch && existingSlot.Commit == commit);
```
Comment (yufeih) [inline src/docfx/restore/index/DependencyGitPool.cs]: `commit` alone is enough to identify a slot, since we are not looking up `branch` from repo.

File: src/docfx/restore/index/DependencySlot.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        public string Id { get; set; }
+
+        public string Url { get; set; }
+
+        public DateTime LastAccessDate { get; set; }
+
+        public bool Restored { get; set; }
+
+        [JsonIgnore]
+        public string Acquirer { get; set; }
```
Comment (yufeih) [inline src/docfx/restore/index/DependencySlot.cs]: Is there a way to transfer acquirer without polluting this data model?

File: src/docfx/restore/index/DependencySlot.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+{
+    internal abstract class DependencySlot
+    {
+        public string Id { get; set; }
+
+        public string Url { get; set; }
+
+        public DateTime LastAccessDate { get; set; }
+
+        public bool Restored { get; set; }
```
Comment (yufeih) [inline src/docfx/restore/index/DependencySlot.cs]: `Commit` is enough to identify if a slot is good

File: src/docfx/restore/index/DependencySlot.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace Microsoft.Docs.Build
+{
+    internal abstract class DependencySlot
+    {
+        public string Id { get; set; }
+
+        public string Url { get; set; }
+
+        public DateTime LastAccessDate { get; set; }
```
Comment (yufeih) [inline src/docfx/restore/index/DependencySlot.cs]: why is there a need to store time?

File: src/docfx/lib/ProcessUtility.cs [hunk 7]
```diff
@@ -275,12 +282,15 @@ public static async Task<string> ReadFile(string path)
...
+        public static Task<T> ReadFile<T>(string path)
+            => ReadFile(path, stream => JsonUtility.Deserialize<T>(new StreamReader(stream).ReadToEnd()));
+
         public static async Task<T> ReadFile<T>(string path, Func<Stream, T> read)
         {
             T result = default;
             await RunInsideMutex(path, () =>
             {
-                using (var fs = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read, 1024, FileOptions.SequentialScan))
+                using (var fs = new FileStream(path, FileMode.OpenOrCreate, FileAccess.Read, FileShare.Read, 1024, FileOptions.SequentialScan))
```
Comment (yufeih) [inline src/docfx/lib/ProcessUtility.cs]: This change the semantic of `ReadFile`, it should not attempt to create a file.

File: src/docfx/lib/ProcessUtility.cs [hunk 8]
```diff
@@ -326,23 +336,20 @@ public static Task WriteFile(string path, Action<Stream> write)
...
         {
             Debug.Assert(!string.IsNullOrEmpty(mutexName));
+            var lockPath = Path.Combine(AppData.MutexRoot, HashUtility.GetMd5Hash(mutexName));
 
-            // avoid the RunInsideMutex to be nested used
-            // doesn't support to require a lock before releasing a lock
-            // which may cause deadlock
-            if (t_innerCall.Value != null)
+            // avoid the RunInsideMutex to be nested used which cause deadlock
+            if (t_innerCall.Value != null && t_innerCall.Value == lockPath)
```
Comment (yufeih) [inline src/docfx/lib/ProcessUtility.cs]: can this be done in another PR? to allow fine grained nest detection? This should be a stack/list to detect deadlocks.

File: src/docfx/restore/lock/DependencyLock.cs [hunk 9]
```diff
@@ -10,21 +10,21 @@ namespace Microsoft.Docs.Build
 {
     internal static class DependencyLock
     {
-        public static DependencyLockModel GetGitLock(this DependencyLockModel dependencyLock, string href, string branch)
+        public static (string key, DependencyLockModel value) GetGitLock(this DependencyLockModel dependencyLock, string href, string branch)
```
Comment (yufeih) [inline src/docfx/restore/lock/DependencyLock.cs]: consider not include this change in this PR.

File: src/docfx/restore/index/DependencyGitPool.cs (hunk 3 above)
Comment (OsmondJiang) [reply src/docfx/restore/index/DependencyGitPool.cs]: not enough, we have `@implicit` feature which based on branch.

File: src/docfx/restore/index/DependencySlot.cs (hunk 6 above)
Comment (OsmondJiang) [reply src/docfx/restore/index/DependencySlot.cs]: it's last access data, used for ordering and apply lockdown time

File: src/docfx/restore/index/DependencySlot.cs (hunk 5 above)
Comment (OsmondJiang) [reply src/docfx/restore/index/DependencySlot.cs]: not enough, i will change the `commit` at the start of restoring.

File: src/docfx/restore/index/DependencySlot.cs (hunk 4 above)
Comment (OsmondJiang) [reply src/docfx/restore/index/DependencySlot.cs]: i would like to keep it with slot. becase slot need to be released anyway.

File: src/docfx/lib/ProcessUtility.cs (hunk 8 above)
Comment (OsmondJiang) [reply src/docfx/lib/ProcessUtility.cs]: ok

File: src/docfx/lib/ProcessUtility.cs (hunk 7 above)
Comment (OsmondJiang) [reply src/docfx/lib/ProcessUtility.cs]: will revert that

File: src/docfx/restore/index/DependencySlotPool.cs (hunk 2 above)
Comment (OsmondJiang) [reply src/docfx/restore/index/DependencySlotPool.cs]: ok

File: src/docfx/build/Build.cs [hunk 10]
```diff
@@ -220,15 +234,16 @@ private static async Task<DependencyLockModel> LoadBuildDependencyLock(string do
             return dependencyLock ?? new DependencyLockModel();
         }
 
-        private static (List<Error> errors, Config config) GetBuildConfig(string docset, Repository repository, CommandLineOptions options, DependencyLockModel dependencyLock)
+        private static (List<Error> errors, Config config) GetBuildConfig(string docset, Repository repository, CommandLineOptions options, DependencyLockModel dependencyLock, DependencyGitPool dependencyGitPool)
         {
             if (ConfigLoader.TryGetConfigPath(docset, out _) || !LocalizationUtility.TryGetSourceRepository(repository, out var sourceRemote, out var sourceBranch, out var locale))
             {
-                return ConfigLoader.Load(docset, options);
+                var (errors, config) = ConfigLoader.Load(docset, options);
```
Comment (OsmondJiang) [inline src/docfx/build/Build.cs]: revert this changes

File: src/docfx/docset/Docset.cs [hunk 11]
```diff
@@ -177,15 +182,15 @@ internal class Docset
                 Debug.Assert(!string.IsNullOrEmpty(Config.Theme));
 
                 var (themeRemote, themeBranch) = LocalizationUtility.GetLocalizedTheme(Config.Theme, Locale, Config.Localization.DefaultLocale);
-                var (themePath, themeLock) = RestoreMap.GetGitRestorePath($"{themeRemote}#{themeBranch}", DependencyLock);
+                var (themePath, themeLock) = _dependencyGitPool.AcquireSharedGit($"{themeRemote}#{themeBranch}", DependencyLock);
```
[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/docset/Docset.cs]: will change back to 'GetGitRestorePath'

File: src/docfx/restore/index/DependencySlotPool.cs (hunk 2 above)
Comment (OsmondJiang) [reply src/docfx/restore/index/DependencySlotPool.cs]: ok

File: src/docfx/restore/lock/DependencyLock.cs (hunk 9 above)
Comment (OsmondJiang) [reply src/docfx/restore/lock/DependencyLock.cs]: i will try to not change the structure

File: src/docfx/restore/RestoreGit.cs [hunk 12]
```diff
@@ -144,20 +148,40 @@ async Task AddWorkTrees()
...
+                                    throw Errors.GitCloneFailed(remote, branches).ToException(ex);
+                                }
                             }
-                            catch (Exception ex)
+                            else
                             {
-                                throw Errors.GitCloneFailed(remote, branches).ToException(ex);
+                                // worktree already exists
+                                // checkout to {headCommit}, no need to fetch
+                                await GitUtility.Checkout(workTreePath, headCommit);
```
Comment (yufeih) [inline src/docfx/restore/RestoreGit.cs]: Is add worktree enough to update `workTreePath` to `headCommit`? since we are calling `worktree add --force`

File: src/docfx/restore/index/DependencySlot.cs [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+using System;
+using Newtonsoft.Json;
+
+namespace Microsoft.Docs.Build
+{
+    internal abstract class DependencySlot
+    {
+        public string Id { get; set; }
+
+        public string Url { get; set; }
```
Comment (yufeih) [inline src/docfx/restore/index/DependencySlot.cs]: Is `Url` just like `Acquirer` for internal tracking purpose? It does not need to persist on disk.

File: src/docfx/restore/index/DependencyGitPool.cs [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.Docs.Build
+{
+    internal class DependencyGitPool
+    {
+        private Dictionary<(string remote, string branch, string commit, LockType lockType), (string path, DependencyGit git, bool released)> _acquiredGits = new Dictionary<(string remote, string branch, string commit, LockType lockType), (string path, DependencyGit git, bool released)>();
```
Comment (yufeih) [inline src/docfx/restore/index/DependencyGitPool.cs]: can this dictionary be `readonly IReadOnlyDictionary`, to separate mutating operations like `AcquireXXX` from non mutating (thread safe) operations like `GetXXX` and `ReleaseXXX`. Like builder patterns used in other places, mutating operations returns an immutable instance.

File: src/docfx/restore/index/DependencySlotPool.cs [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+
+                    WriteSlots(restoreDir, slots);
+                });
+
+            Debug.Assert(released);
+
+            return released;
+        }
+
+        public static async Task<(string path, T slot)> AcquireSlots<T>(string url, LockType type, Func<T, T> updateExistingSlot, Func<T, bool> matchExistingSlot) where T : DependencySlot, new()
```
Comment (mingwli) [inline src/docfx/restore/index/DependencySlotPool.cs]:
```suggestion
        public static async Task<(string path, T slot)> AcquireSlot<T>(string url, LockType type, Func<T, T> updateExistingSlot, Func<T, bool> matchExistingSlot) where T : DependencySlot, new()
```

File: src/docfx/restore/index/DependencySlot.cs (hunk 13 above)
Comment (OsmondJiang) [reply src/docfx/restore/index/DependencySlot.cs]: you r right.

File: src/docfx/restore/index/DependencyGitPool.cs (hunk 14 above)
Comment (OsmondJiang) [reply src/docfx/restore/index/DependencyGitPool.cs]: great idea, try try

File: src/docfx/restore/index/DependencySlotPool.cs (hunk 15 above)
Comment (OsmondJiang) [reply src/docfx/restore/index/DependencySlotPool.cs]: 赞！

File: src/docfx/restore/RestoreGit.cs (hunk 12 above)
Comment (OsmondJiang) [reply src/docfx/restore/RestoreGit.cs]: no, it will throw work tree already exists error, because the work tree name is not changed

File: src/docfx/restore/RestoreMap.cs [hunk 16]
```diff
@@ -12,100 +12,8 @@ namespace Microsoft.Docs.Build
 {
     internal static class RestoreMap
```
Comment (mingwli) [inline src/docfx/restore/RestoreMap.cs]: >RestoreMap [](start = 26, length = 10)

Only restore-file-related logic in this class, would it be better to merge it with RestoreFile?

Comment (OsmondJiang) [reply src/docfx/restore/RestoreMap.cs]: i will change `restore file` part in another pr

File: src/docfx/build/Build.cs [hunk 17]
```diff
@@ -15,22 +15,35 @@ internal static class Build
...
-            XrefMap xrefMap = null;
-
             var repository = Repository.Create(docsetPath);
             Telemetry.SetRepository(repository?.Remote, repository?.Branch);
 
             var locale = LocalizationUtility.GetLocale(repository, options);
             var dependencyLock = await LoadBuildDependencyLock(docsetPath, locale, repository, options);
-            var (configErrors, config) = GetBuildConfig(docsetPath, repository, options, dependencyLock);
+
+            var dependencyGitPool = await DependencyGitPool.AcquireGitPool(dependencyLock);
```
Comment (live1206) [inline src/docfx/build/Build.cs]: Does `AcquireGitPool` mean `Create`?

File: src/docfx/restore/index/DependencyGitPool.cs [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+            catch
+            {
+                successed = false;
+                throw;
+            }
+            finally
+            {
+                if (!successed && root)
+                {
+                    foreach (var (k, v) in acquired)
```
Comment (live1206) [inline src/docfx/restore/index/DependencyGitPool.cs]: Can we just call `Release` here?
Since `AcquireGitPool` is supposed to be invoked only once to get the instance?

Comment (OsmondJiang) [reply src/docfx/restore/index/DependencyGitPool.cs]: offline discussed

File: src/docfx/lib/ProcessUtility.cs [hunk 19]
```diff
@@ -22,6 +22,25 @@ internal static class ProcessUtility
         private const int _defaultLockExpireTimeInSecond = 60 * 60 * 6; /*six hours*/
         private static AsyncLocal<ImmutableStack<string>> t_innerCall = new AsyncLocal<ImmutableStack<string>>();
 
+        public static async Task<bool> IsExclusiveLockHeld(string lockName)
+        {
```
Comment (live1206) [inline src/docfx/lib/ProcessUtility.cs]: update UT to check this?

File: src/docfx/restore/index/DependencySlotPool.cs [hunk 20]
```diff
@@ -0,0 +1,219 @@
...
+            var released = false;
+            await ProcessUtility.RunInsideMutex(
+                url + "/index.json",
+                async () =>
+                {
+                    var slots = GetSlots<T>(restoreDir);
+                    var slotsToRelease = slots.Where(i => i.Id == slot.Id);
+                    Debug.Assert(slotsToRelease.Count() == 1);
+
+                    var slotToRelease = slotsToRelease.First();
```
Comment (live1206) [inline src/docfx/restore/index/DependencySlotPool.cs]: use `slotsToRelease.Single()` to get rid of `Debug.Assert` above and below

File: src/docfx/restore/index/DependencySlotPool.cs [hunk 21]
```diff
@@ -0,0 +1,219 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.Docs.Build
+{
+    internal static class DependencySlotPool
+    {
+        private static int _defaultLockdownTimeInSecond = 10 * 60;
+
+        public static async Task<(string path, T slot)> TryGetSlot<T>(string remote, Func<IReadOnlyList<T>, IReadOnlyList<T>> matchExistingSlots) where T : DependencySlot
```
Comment (live1206) [inline src/docfx/restore/index/DependencySlotPool.cs]: possible to put `<T>` on class level to get rid of `where` constraint on every method?
if not, do we need where `T : new ()` here?

Comment (OsmondJiang) [reply src/docfx/restore/index/DependencySlotPool.cs]: we need`T :  new()` here to new a T.
